### ISSUE TITLE
1.5GB of Eyre channels, most things now cause bail: meme

### ISSUE THREAD

**Issue by forfel-norfel** [urbit/urbit#3051]
I don't know what caused this.

```
~forfel-norfel:dojo> |mass/
%userspace:      
  %hoon: MB/4.381.676
  %zuse: MB/5.337.328
  %caches:              
    %vane-a: MB/5.286.192         
    %vane-b: KB/332.136         
    %vane-c: MB/4.541.796    
    %vane-d: KB/684.700           
    %vane-e: MB/1.480.176     
    %vane-f: MB/4.201.668
    %vane-g: MB/2.209.728      
    %vane-i: KB/203.940          
    %vane-j: KB/947.980         
  --MB/19.888.316      
  %vanes:                         
    %dill:                    
      %hey: B/96    
      %dug: KB/1.632             
      %dot: B/416               
    --KB/2.144          
    %ames:                    
      %peers-known: MB/4.763.180
      %peers-alien: KB/38.124   
      %dot: MB/2.110.884  
    --MB/6.912.188         
    %behn:                           
      %timers: KB/258.120           
      %dot: B/272              
    --KB/258.392                 
    %clay:                      
      %domestic: MB/53.637.068     
      %foreign: MB/72.169.076           
      %object-store:              
        %commits: KB/390.572         
        %blobs: MB/80.954.992
      --MB/81.345.564       
      %dot: B/600          
    --MB/207.152.308          
    %eyre:                                                                                                                                                                                   
      %bindings: KB/5.044                                                                                                                                                                    
      %auth: KB/4.056                                                                                                                                                                        
      %connections: B/0                                                                                                                                                                      
      %channels: GB/1.553.702.248
      %axle: KB/9.376             
    --GB/1.553.720.724             
    %ford:                              
      %builds: MB/49.060.420
      %compiler-cache: B/264
      %dot: KB/528.564
    --MB/49.589.248
    %gall:                    
      %foreign: KB/14.760  
      %blocked:  
      --B/0          
      %active:       
        %acme: KB/31.512
        %azimuth-tracker: KB/1.388
        %chat-cli: MB/16.881.024
        %chat-hook: KB/56.360
        %chat-store: MB/49.840.388
        %chat-view: KB/104.208
        %clock: KB/1.020 
        %contact-hook: KB/6.788
        %contact-store: KB/25.392
        %contact-view: KB/86.088
        %dojo: KB/5.296
        %eth-watcher: MB/4.420.912
        %file-server: KB/2.368
        %goad: B/924
        %group-hook: KB/3.400    
        %group-store: KB/120.928
        %hood: KB/18.012
        %invite-hook: KB/1.092
        %invite-store: KB/5.676 
        %invite-view: KB/187.532
        %launch: KB/95.156
        %lens: B/780       
        %link-listen-hook: KB/405.940                                                                                                                                                        
        %link-proxy-hook: KB/329.916                                                                                                                                                         
        %link-store: KB/118.652                                                                                                                                                              
        %link-view: KB/14.856                                                                                                                                                                
        %metadata-hook: KB/2.400                                                                                                                                                             
        %metadata-store: KB/170.848                                                                                                                                                          
        %permission-group-hook: KB/1.932                                                                                                                                                     
        %permission-hook: KB/3.516                                                                                                                                                           
        %permission-store: KB/111.664                                                                                                                                                        
        %ping: KB/1.356                                                                                                                                                                      
        %publish: KB/153.892                                                                                                                                                                 
        %s3-store: KB/1.844                                                                                                                                                                  
        %soto: B/804                                                                                                                                                                         
        %spider: KB/236.648                                                                                                                                                                  
        %weather: KB/55.976                                                                                                                                                                  
      --MB/73.506.488                                                                                                                                                                        
      %dot: KB/121.744                                                                                                                                                                       
    --MB/73.642.992              
    %iris:                        
      %nex: B/0                    
      %outbound: B/160                  
      %by-id: B/0           
      %by-duct: B/0         
      %axle: B/96     
    --B/256        
    %jael:                    
      %pki: MB/1.591.240   
      %etn: B/168
      %dot: B/48     
    --MB/1.591.456   
  --GB/1.892.869.708    
  %vases:                         
    %vane-a: B/312              
    %vane-b: B/192           
    %vane-c: B/216                
    %vane-d: B/264            
    %vane-e: B/192       
    %vane-f: B/192             
    %vane-g: KB/1.980            
    %vane-i: B/192              
    %vane-j: B/192     
  --KB/3.732       
  %dot: KB/12.972             
--GB/1.922.493.732         
total userspace: GB/1.922.493.732
  kernel: B/24       
  wish cache: B/120  
total arvo stuff: B/144 
  warm jet state: KB/79.980       
  cold jet state: MB/114.350.972
  hank cache: B/372          
  battery hash cache: B/288       
  call site cache: B/100      
  hot jet state: KB/128.224
total jet stuff: MB/114.559.936
  bytecode programs: MB/1.766.272
  bytecode cache: KB/157.100    
total nock stuff: MB/1.923.372
  trace buffer: B/36              
  memoization cache: KB/1.296 
total road stuff: KB/1.332
space profile: KB/2.232          
event: B/144                    
effects: KB/2.976       
total marked: GB/2.038.983.868
free lists: MB/50.946.516       
sweep: GB/2.038.983.868         
> |mass                   
>=                         
~forfel-norfel:dojo>
```

**Comment by joemfb** [comment]
I ran into this as well. Those channels are persistent (timing out after `~h12`), and buffer output. I think there are some pathological retry cases in the interface that can request massive amounts of data (for instance, full backlog for long-lived chat channels). Luckily, the space will be reclaimed after the 12 hour timeout, so you'll just need to keep your ship idle until then. I'd just shut it down.

After the channels expire, your persistent state (the "home road" of the "loom") will be much smaller, but likely still fairly fragmented. You'll want to try to deduplicate and compact it with `|pack` (which doesn't work in all cases, but your state should be small enough).

Better management of persistent state (and response to memory pressure) is very high on our priority list. This (relatively rare) runaway space-leak in %eyre is particularly unacceptable.

**Comment by ohAitch** [comment]
Does this mean chat on load receives 1.5GB of JSON in sufficiently large
channels, instead of requesting messages incrementally?

**Comment by joemfb** [comment]
No, it means that %eyre has (or had) open gall subscriptions across all of those channels which resulted in 1.5GB of serialized responses, which are persistently buffered until either a) acknowledged by the front-end or b) channel timeout. (Most of those serialized responses are probably json, and most of the subscriptions are probably for chat backlog.)

**Comment by ohAitch** [comment]
Ah, did the "after 10 unacked responses unsubscribe" logic in gall die
then? (I forget the exact number, but there definitely used to be the
backpressure mechanism you describe)

**Comment by philipcmonk** [comment]
Yes, that's gone and is now handled by each client vane. For example, Ames'
rule is >5 unacked messages and no contact for 30 seconds

**Comment by ohAitch** [comment]
Or (virtually) unhandled, in this case P:

Contact on that particular channel, or contact period? Could another ship
hypothetically have a hung channel client causing bufferbloat, as long as
some unrelated channel is running?

((Is the motivation for having subscription timeout be 12 hours something
along the lines of, the web chat client / chat subscription protocol
doesn't implement reconnects after partial data transmission?))

**Comment by xiphiness** [comment]
```  %zuse: MB/2.938.456
  %vane: 
    %reports: 
      %vane-a: 
        %peers-known: MB/14.913.848
        %peers-alien: KB/488.044
      --MB/15.401.892
      %vane-b: 
        %timers: MB/1.196.936
      --MB/1.196.936
      %vane-c: 
        %domestic: 
          %base: 
            %ankh: MB/25.806.320
            %mime: MB/3.692.916
            %ford-vases: B/0
            %ford-marks: B/0
            %ford-casts: B/0
          --MB/29.499.236
          %home: 
            %ankh: MB/2.062.312
            %mime: KB/86.332
            %ford-vases: MB/25.965.596
            %ford-marks: MB/138.356.624
            %ford-casts: KB/22.632
          --MB/166.493.496
          %kids: 
            %ankh: B/0
            %mime: B/0
            %ford-vases: B/0
            %ford-marks: B/0
            %ford-casts: B/0
          --B/0
        --MB/195.992.732
        %foreign: MB/7.597.120
        %object-store: 
          %commits: KB/686.848
          %blobs: B/0
        --KB/686.848
      --MB/204.276.700
      %vane-d: 
        %hey: B/24
        %dug: KB/1.632
      --KB/1.656
      %vane-e: 
        %bindings: KB/3.724
        %auth: KB/3.132
        %connections: B/0
        %channels: GB/1.059.320.872
        %axle: KB/9.412
      --GB/1.059.337.140
      %vane-f: 
      --B/0
      %vane-g: 
        %foreign: KB/87.148
        %blocked: 
        --B/0
        %active: 
          %acme: KB/12.940
          %azimuth-tracker: KB/1.544
          %chat-cli: MB/53.132.528
          %chat-hook: KB/402.064
          %chat-store: MB/162.724.200
          %chat-view: KB/471.824
          %clock: B/984
          %contact-hook: KB/59.448
          %contact-store: KB/108.736
          %contact-view: KB/420.984
          %dojo: KB/3.684
          %eth-watcher: MB/5.021.072
          %file-server: MB/12.767.532
          %glob: KB/1.164
          %goad: B/912
          %graph-store: B/816
          %group-hook: KB/2.328
          %group-pull-hook: KB/6.648
          %group-push-hook: KB/2.088
          %group-store: KB/366.860
          %hood: KB/23.568
          %invite-hook: KB/1.020
          %invite-store: KB/133.508
          %invite-view: KB/620.036
          %launch: KB/448.080
          %lens: B/816
          %link-listen-hook: MB/1.121.572
          %link-proxy-hook: KB/703.964
          %link-store: KB/309.292
          %link-view: KB/34.496
          %metadata-hook: KB/8.796
          %metadata-store: KB/912.020
          %permission-group-hook: KB/3.168
          %permission-hook: KB/12.140
          %permission-store: KB/438.456
          %ping: KB/1.368
          %publish: KB/58.224
          %s3-store: KB/260.264
          %soto: B/696
          %spider: MB/1.712.448
          %weather: KB/179.544
        --MB/242.491.832
      --MB/242.578.980
      %vane-i: 
        %nex: B/0
        %outbound: B/160
        %by-id: B/588
        %by-duct: B/72
        %axle: B/132
      --B/952
      %vane-j: 
        %pki: MB/1.786.256
        %etn: KB/9.936
      --MB/1.796.192
    --GB/1.524.590.448
    %caches: 
      %vane-a: MB/2.280.400
      %vane-b: KB/323.256
      %vane-c: KB/56.256
      %vane-d: KB/435.320
      %vane-e: MB/1.547.168
      %vane-f: KB/14.448
      %vane-g: KB/447.760
      %vane-i: KB/184.504
      %vane-j: KB/867.380
    --MB/6.156.492
    %dregs: 
      %vane-a: KB/115.968
      %vane-b: B/440
      %vane-c: KB/20.332
      %vane-d: B/336
      %vane-e: B/192
      %vane-f: B/120
      %vane-g: KB/833.136
      %vane-i: B/192
      %vane-j: B/240
    --KB/970.956
  --GB/1.531.717.896
--GB/1.539.052.992
total userspace: GB/1.539.052.992
  kernel: KB/102.132
total arvo stuff: KB/102.132
  warm jet state: KB/92.320
  cold jet state: MB/78.911.472
  hank cache: B/372
  battery hash cache: B/288
  call site cache: B/100
  hot jet state: KB/125.152
total jet stuff: MB/79.129.704
  bytecode programs: MB/1.377.236
  bytecode cache: KB/126.588
total nock stuff: MB/1.503.824
  trace buffer: B/36
  memoization cache: B/288
total road stuff: B/324
space profile: KB/8.724
total marked: GB/1.619.797.700
free lists: MB/15.595.696
sweep: GB/1.619.797.700````


Seem to be getting this issue well after 12 hours - if I'm online my ship lasts a few minutes before bail memeing, even with urbit-worker deduplication, so I have had it running in offline (-L) mode in hopes that the eyre channels eventually get kicked, to no avail.

**Comment by botter-nidnul** [comment]
I've had the same happen to me once @xiphiness  and `|reload %e` fixed the problem.

**Comment by xiphiness** [comment]
Ah, thanks for that! I should have thought of that :) That did help a bit - it brought the cache from `%vane-e: MB/1.547.168` to `%vane-e: MB/2.229.340` but channels is still untouched at `%channels: GB/1.059.320.824`.

**Comment by galenwp** [comment]
I would love to track the root cause of this down

**Comment by matildepark** [comment]
There seems to be a bunch of related (symptomatic?) bugs around this, yeah.

**Comment by xiphiness** [comment]
Yeah, 15 hours later still ´%channels: GB/1.059.320.824´

Maybe a short fix would be a note to send to eyre from a hood generator(is that what they are called?) that clears all channels etc when this problem arises.

**Comment by joemfb** [comment]
You'll need for those to be completely inactive to hit the ~12hr timeout (ie, you'd have to shut your ship down).

But there already is a move to force-clear these. Once #3472 lands, you'll be able to trigger it easily with `|trim` (or `|trim 0`), but, for now, it's a bit of a manual process. You'll need to manually construct the trim event `.trim/jam [//arvo trim+1]` (or `.trim/jam [//arvo trim+0]`, but there's no difference as far as %eyre is concerned), and then restart the process, injecting that event:

```
$ ./path/to/urbit -I path/to/pier/.urb/put/trim.jam path/to/pier
~
urbit 0.10.8
...
pier: injecting %trim event on //arvo
...
```

%eyre only clears inactive channels on `%trim`, so you may need to close all relevant browser tabs first.

(Long-term, my view is that the %eyre channel buffering needs to be capped in some way. And we need to get subscription lifecycles under control in userspace and the frontend.)

**Comment by xiphiness** [comment]
oh it just needs to be off for 12 hours? I thought I needed to not connect to landscape for 12 hours while it was on.

so I can

´echo ".trim/jam [//arvo trim+1]" > path/to/pier/.urb/put/trim.jam´

to create the jam file?

curious, what is the difference between the 1 and the 0, even if it has no impact on eyre, and what exactly is trim doing? my impression is that it might ping vanes to clear any stores they don't fundamentally need, but I might be mistaken

thanks! very useful

**Comment by belisarius222** [comment]
Shouldn't there be some way we could store a client-id of some kind that remains stable across reloads, so that when that client reconnects, Eyre would delete channels related to previous instances of that channel id from its state?  I guess it would be one client-id per browser tab, ideally.

Most people only have one urbit tab open at a time, so this proliferation of Eyre channels seems like it should be avoidable.

An alternative approach could be to try to make the data structures stored in Eyre easier for Vere to deduplicate, like what was done in Ames a few months ago.

—
~rovnys-ricfer
https://urbit.org

**Comment by joemfb** [comment]
@belisarius222, there are definitely options, they just need to be designed/analyzed.

@xiphiness, the trim number is a priority: `0` is highest priority and `1` is lower priority. This was chosen for future extensibility, we'll probably have even lower priorities as we gain more granular control. `%eyre` trims in the same manner regardless of the priority (and only trims inactive channels, which is probably why you had to run it several times).

**Comment by xiphiness** [comment]
ahh ok

got back up to 1.5GB again in about 2 days and bail memed.... maybe this is my poor tab usage, opening an extra one because I don't see the other in my sea of hundreds of tabs, also opening on my phone when I'm out for a smoke, etc

**Comment by joemfb** [comment]
Note that this issue could be caused (or exacerbated) by `channel.js` only sending acknowledgements on write `sendJSONToChannel()` -- read-only subscriptions appear to never acknowledge their results.

/cc @baudtack @eglaysher  @tacryt-socryp @liam-fitzgerald

**Comment by matildepark** [comment]
Is this related or unrelated to https://github.com/urbit/landscape/issues/40 ? We have @baudtack on that.

**Comment by joemfb** [comment]
It's the acknowledgements in the opposite direction, ie, the client informing %eyre that it's received everything up to message N, as opposed to %eyre informing the client that it has received a request.

**Comment by joemfb** [comment]
While I'm at it, `channel.js` is stored in the repository as a binary file, so it has no line-by-line diff/blame/history &c.

**Comment by matildepark** [comment]
Wait, [doesn't it](https://github.com/urbit/urbit/blame/master/pkg/arvo/app/landscape/js/channel.js)?

**Comment by joemfb** [comment]
Ok, looks like only the CLI is affected. The cause is `pkg/arvo/**/*.js binary` in the top-level `.gitattributes`.

**Comment by matildepark** [comment]
@Fang- Can we close this (and possibly other related ones)?

**Comment by Fang-** [comment]
Yes, I think such cases shouldn't pop up anymore now.
